pkg/storage/cacher.go

```diff
@@ -355,9 +357,24 @@ func (c *Cacher) Codec() runtime.Codec {
 func (c *Cacher) processEvent(event watchCacheEvent) {
 	c.Lock()
 	defer c.Unlock()
-	for _, watcher := range c.watchers {
-		watcher.add(event)
-	}
+
+	// prevent creating too many goroutines
+	burstChan := make(chan struct{}, burstWatchers)
+	// prevent sending events to a watcher out of order
+	wait := sync.WaitGroup{}
+
+	wait.Add(len(c.watchers))
+	for _, w := range c.watchers {
+		burstChan <- struct{}{}
+		go func(watcher *cacheWatcher) {
+			defer func() {
+				<-burstChan
+				wait.Done()
+			}()
+			watcher.add(event)
+		}(w)
+	}
+	wait.Wait()
```

Comment:
I think that having bigger buffers may be better, because you avoid the synchronization overhead that you are doing in this PR. Although your PR looks generally great, but it may slow down system in a healthy situation